pkg/keyring/keyring_test.go

```diff
@@ -0,0 +1,115 @@
+package keyring
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	"github.com/smartcontractkit/chainlink-relay/pkg/keyring/algo"
+	"github.com/smartcontractkit/chainlink-relay/pkg/keyring/backend/sql"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+)
+
+func getCodec() codec.Codec {
+	registry := codectypes.NewInterfaceRegistry()
+	cryptocodec.RegisterInterfaces(registry)
+	return codec.NewProtoCodec(registry)
+}
+
+var algoOptionFn = func(options *keyring.Options) {
+	options.SupportedAlgos = keyring.SigningAlgoList{
+		algo.Secp256k1,
+		algo.Ed25519,
+	}
+
+	options.SupportedAlgosLedger = keyring.SigningAlgoList{
+		algo.Secp256k1,
+		algo.Ed25519,
+	}
+}
+
+func newKeyringInMem() keyring.Keyring {
+	return keyring.NewInMemory(getCodec(), algoOptionFn)
+}
+
+func newKeyringSql() keyring.Keyring {
+	backend := sql.NewKeyring(nil)
+	return keyring.NewKeystore(backend, getCodec(), algoOptionFn)
+}
+
+func TestNewService(t *testing.T) {
+	keyrings := Keyrings{
+		"ethereum.tx-ocr_report.test/mem": newKeyringInMem(),
+		"ethereum.tx-ocr_report.test/sql": newKeyringSql(),
+		"solana.tx.test/mem":              newKeyringInMem(),
+		"solana.ocr-report.test/mem":      newKeyringInMem(),
+		"solana.tx.test/sql":              newKeyringSql(),
+		"solana.ocr-report.test/sql":      newKeyringSql(),
+	}
+
+	s := NewService(nil, keyrings)
+	assert.NotEmpty(t, s)
+	assert.NotEmpty(t, s.Signers["ethereum.tx-ocr_report.test/mem"])
+	assert.NotEmpty(t, s.Signers["ethereum.tx-ocr_report.test/sql"])
+	assert.NotEmpty(t, s.Signers["solana.tx.test/mem"])
+	assert.NotEmpty(t, s.Signers["solana.ocr-report.test/mem"])
+}
+
+func TestNewServicePortsSignVerify(t *testing.T) {
+	keyringEthTest := "ethereum.tx.test/mem"
+	keyrings := Keyrings{
+		"ethereum.tx.test/mem": newKeyringInMem(),
+	}
+
+	// generate a key to test
+	mnemonic := "health amateur need boy enough bless april march dove rabbit satoshi purse"
+	bip39Passphrase := ""
+	hdPath := "m/44'/60'/0'/0/0"
+
+	testCases := []struct {
+		name string
+		uid  string
+		algo keyring.SignatureAlgo
+	}{
+		{
+			name: "can sign/verify secp256k1",
+			uid:  "uuid-secp256k1-1",
+			algo: algo.Secp256k1,
+		},
+		{
+			name: "can sign/verify ed25519",
```

Comment:
Also note: for verify operations https://github.com/hdevalence/ed25519consensus needs to be used. It only implements `Verify` and works well together with `crypto/ed25519` https://github.com/hdevalence/ed25519consensus/blob/main/ed25519_test.go#L12-L18